game-core/src/main/java/games/strategy/triplea/delegate/battle/subs/Subs.java

```diff
@@ -0,0 +1,91 @@
+package games.strategy.triplea.delegate.battle.subs;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.MustFightBattle.ReturnFire;
+import java.util.Collection;
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class Subs {
+
+  public enum FireOrder {
+    DEF_BEFORE_ATT,
+    DEF_BEFORE_REGULAR,
+    DEF_WITH_REGULAR
+  }
+
+  public static FireOrder getFireOrder(
+      final Collection<Unit> attackingUnits,
+      final Collection<Unit> defendingUnits,
+      final GameData gameData) {
+    if (defenderSubsFireFirst(attackingUnits, defendingUnits, gameData)) {
+      return FireOrder.DEF_BEFORE_ATT;
+    }
+    final boolean defendingSubsFireWithAllDefenders =
+        !Properties.getWW2V2(gameData)
+            && returnFireAgainstDefendingSubs(attackingUnits, defendingUnits, gameData)
+                == ReturnFire.ALL;
+    if (defendingSubsSneakAttack(gameData) && !defendingSubsFireWithAllDefenders) {
+      return FireOrder.DEF_BEFORE_REGULAR;
+    }
+    return FireOrder.DEF_WITH_REGULAR;
+  }
+
+  public static boolean defenderSubsFireFirst(
+      final Collection<Unit> attackingUnits,
+      final Collection<Unit> defendingUnits,
+      final GameData gameData) {
+    return returnFireAgainstAttackingSubs(attackingUnits, defendingUnits, gameData)
+            == ReturnFire.ALL
+        && returnFireAgainstDefendingSubs(attackingUnits, defendingUnits, gameData)
+            == ReturnFire.NONE;
+  }
+
+  public static ReturnFire returnFireAgainstAttackingSubs(
+      final Collection<Unit> attackingUnits,
+      final Collection<Unit> defendingUnits,
+      final GameData gameData) {
```

Comment:
What do you think of making GameData the first param in these cases? It seems a bit more natural to me.

Reply:
I had noticed that many of the existing methods have GameData as the last parameter.  So I was following that convention.  But I can move it to the first parameter.  I don't have an opinion either way.